src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/RecoverReadTags.java

```diff
@@ -0,0 +1,154 @@
+package org.broadinstitute.hellbender.tools.walkers.qc;
+
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.util.PeekableIterator;
+import org.broadinstitute.barclay.argparser.Argument;
+import org.broadinstitute.barclay.argparser.CommandLineProgramProperties;
+import org.broadinstitute.hellbender.cmdline.StandardArgumentDefinitions;
+import org.broadinstitute.hellbender.cmdline.programgroups.QCProgramGroup;
+import org.broadinstitute.hellbender.cmdline.programgroups.ShortVariantDiscoveryProgramGroup;
+import org.broadinstitute.hellbender.engine.GATKPath;
+import org.broadinstitute.hellbender.engine.GATKTool;
+import org.broadinstitute.hellbender.engine.ReadsDataSource;
+import org.broadinstitute.hellbender.engine.ReadsPathDataSource;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadQueryNameComparator;
+import org.broadinstitute.hellbender.utils.read.SAMFileGATKReadWriter;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * This tool is designed for a pair of SAM files sharing the same read names (e.g.
+ * SAM files before and after alignment), where one SAM lacks some read tags that are
+ * present in the other.
+ *
+ * This situation may happen if for instance an unaligned bam is converted to fastq files,
+ * and some read tags get lost during this conversion. The converting to fastq is sometimes unavoidable,
+ * due to the fact that some tools, such as the adapter clipping tools, are written for fastq only,
+ * while others are written specifically for SAM files.
+ *
+ * This tools behaves similarly to Picard MergeBamAlignment (MBA). The difference is that whereas
+ * MBA merges the alignment information to the unaligned bam, RecoverReadTags uses the aligned bam as the
+ * base and adds the read tags from the unaligned bam to the aligned bam.
+ *
+ * Currently, the tool is implemented for the specific case of recovering UMI read tags (RX)
+ * from an unaligned bam.
+ */
+@CommandLineProgramProperties(
+        summary = "Incorporate read tags in a SAM file to that of a matching SAM file",
+        oneLineSummary = "Incorporate read tags in a SAM file to that of a matching SAM file",
+        programGroup = QCProgramGroup.class // sato: change
+)
+public class RecoverReadTags extends GATKTool {
+    @Argument(fullName = "unmapped-sam", doc = "query-name sorted unmapped sam file containing the read tag of interest")
+    public GATKPath unmappedSamFile;
+    ReadsDataSource unmappedSam;
+
+    @Argument(shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, doc = "")
+    public File outSamFile;
+    SAMFileGATKReadWriter writer;
+
+    final static String UMI_TAG = "RX";
+    final static List<String> DEFAULT_READ_TAGS = Arrays.asList(UMI_TAG);
```

Comment:
```suggestion
        final static List<String> DEFAULT_READ_TAGS = new ArrayList() {{
        add(UMI_TAG);
    }};
```

Reply:
@takutosato The list returned by Arrays.asList() is immutable, so the parser can't modify the `DEFAULT_READ_TAGS` list. See the suggested alternative above.

Reply:
I was getting errors (warnings, really) related to serializable UID stuff, which I don't understand well. The simpler approach seems to be to just have an empty list be the default, so that's what I did in the latest commit.